istioctl/cmd/analyze.go

```diff
@@ -0,0 +1,103 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"istio.io/istio/galley/pkg/config/analysis/analyzers"
+	"istio.io/istio/galley/pkg/config/analysis/local"
+	"istio.io/istio/galley/pkg/config/processor/metadata"
+	"istio.io/istio/galley/pkg/source/kube/client"
+	"istio.io/pkg/log"
+
+	"github.com/spf13/cobra"
+)
+
+// Analyze command
+func Analyze() *cobra.Command {
+	analysisCmd := &cobra.Command{
+		Use:   "analyze <file|globpattern>...",
+		Short: "Analyze Istio configuration and print validation messages",
+		Example: `
+# Analyze yaml files
+istioctl experimental analyze a.yaml b.yaml
+
+# Analyze the current live cluster
+istioctl experimental analyze -c $HOME/.kube/config
+
+# Analyze the current live cluster, simulating the effect of applying additional yaml files
+istioctl experimental analyze -c $HOME/.kube/config a.yaml b.yaml
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			loggingOptions.SetOutputLevel("processing", log.ErrorLevel)
+			loggingOptions.SetOutputLevel("source", log.ErrorLevel)
+			if err := log.Configure(loggingOptions); err != nil {
+				return err
+			}
+
+			files, err := gatherFiles(args)
+			if err != nil {
+				return err
+			}
+			cancel := make(chan struct{})
+
+			sa := local.NewSourceAnalyzer(metadata.MustGet(), analyzers.All())
+
+			// If we're using kube, use that as a base source.
+			if kubeconfig != "" {
+				k, err := client.NewKubeFromConfigFile(kubeconfig)
+				if err != nil {
+					return err
+				}
+				sa.AddRunningKubeSource(k)
+			}
```

Comment:
I don't think the use of `--kubeconfig` to access the cluster vs. offline mode is consistent with other commands. In general, if `--kubeconfig` is not specified, we'll use the defaults (e.g. `$HOME/.kube/config`) rather than falling back to offline mode. 

A few alternatives to consider:

1) Add an explicit option to enable/disable cluster as a source, e.g. default to access current cluster with an option to force offline mode with `--offline`, e.g.

```bash
istioctl x analyze # cluster as source
istioctl x analyze --offline -f a.yaml # offline only
istioctl x analyze a.yaml # file + cluster as source
```

2) Remove cluster as a source and use input files (inc. STDIN) as the only source. Users would need to `kubectl get` resources and pipe to istioctl, e.g.

```bash
# all Istio CRD
kubectl get istio-io --all-namespaces | istioctl x analyze -f -

# CRD for my service (namespace)
kubectl get -n <my-service> | istioctl analyze -f -
```

Reply:
I actually had a separate arg to enable kube originally, but removed it after [earlier feedback](https://github.com/istio/istio/pull/16595#discussion_r318417283) from @liamawhite. 

> In general, if --kubeconfig is not specified, we'll use the defaults (e.g. $HOME/.kube/config) rather than falling back to offline mode.

That's what I originally had expected, but it doesn't seem to currently be the case in `istioctl`. The --kubeconfig arg defined in `istioctl/cmd/root.go` is blank by default. So even if you define a separate flag, you would still *also* need to supply a --kubeconfig arg. (Unless, of course, we change the behavior of the --kubeconfig flag in `istioctl` more broadly, which is out of scope for what we're currently trying to do). 


Option 2 isn't particularly user friendly IMHO. Istioctl analyze already has commands that can talk to the API server, why make users do it manually in this case?

Reply:
istioctl is using `istio.io/istio/pkg/kube` for building the client config. See below for the loading rules.

https://github.com/istio/istio/blob/0699bc6c7e5b7660ba3d58e1639adc861c1fc549/pkg/kube/config.go#L41-L45

Reply:
Agreed that some form of option (1) is simpler. Option (2) could be added later by accepting stdin as a valid filename (i.e. `-`).

Reply:
Got it, easy fix to use that. Re-added the --use-kube flag.